File: ios/UmbraCore/Sources/UmbraCore/Net/IPv4Packet.swift
import Foundation

/// Minimal IPv4: parse the header to find UDP, and build a reply packet with
/// correct header + UDP checksums. Mirrors the Android Ipv4Packet.
public struct IPv4Packet {
    public static let protoUDP = 17

    public let raw: [UInt8]
    public let length: Int
    public let ihl: Int
    public let proto: Int
    public let srcAddr: [UInt8]
    public let dstAddr: [UInt8]

    public var payloadOffset: Int { ihl }
    public var payloadLength: Int { length - ihl }

    public static func parse(_ buf: [UInt8], _ length: Int) -> IPv4Packet? {
        if length < 20 { return nil }
        let version = (Int(buf[0]) & 0xF0) >> 4
        if version != 4 { return nil }
        let ihl = (Int(buf[0]) & 0x0F) * 4
        if ihl < 20 || ihl > length { return nil }
        let proto = Int(buf[9])
        return IPv4Packet(raw: buf, length: length, ihl: ihl, proto: proto,
                          srcAddr: Array(buf[12..<16]), dstAddr: Array(buf[16..<20]))
    }

    /// Assemble a full IPv4 + UDP packet with recomputed checksums.
    public static func buildUDP(src: [UInt8], dst: [UInt8],
                                srcPort: Int, dstPort: Int, payload: [UInt8]) -> [UInt8] {
        let udpLen = 8 + payload.count
        let total = 20 + udpLen
        var p = [UInt8](repeating: 0, count: total)

        p[0] = 0x45
        p[2] = UInt8((total >> 8) & 0xFF)
        p[3] = UInt8(total & 0xFF)
        p[8] = 64
        p[9] = UInt8(protoUDP)
        p[12] = src[0]; p[13] = src[1]; p[14] = src[2]; p[15] = src[3]
        p[16] = dst[0]; p[17] = dst[1]; p[18] = dst[2]; p[19] = dst[3]
        let ipChk = InternetChecksum.compute(p, 0, 20)
        p[10] = UInt8((ipChk >> 8) & 0xFF)
        p[11] = UInt8(ipChk & 0xFF)

        let u = 20
        p[u] = UInt8((srcPort >> 8) & 0xFF); p[u + 1] = UInt8(srcPort & 0xFF)
        p[u + 2] = UInt8((dstPort >> 8) & 0xFF); p[u + 3] = UInt8(dstPort & 0xFF)
        p[u + 4] = UInt8((udpLen >> 8) & 0xFF); p[u + 5] = UInt8(udpLen & 0xFF)
        for (k, b) in payload.enumerated() { p[u + 8 + k] = b }

        let seed = InternetChecksum.partialSum(src, 0, 4)
            + InternetChecksum.partialSum(dst, 0, 4)
            + UInt64(protoUDP) + UInt64(udpLen)
        var udpChk = InternetChecksum.compute(p, u, udpLen, seed: seed)
        if udpChk == 0 { udpChk = 0xFFFF }
        p[u + 6] = UInt8((udpChk >> 8) & 0xFF)
        p[u + 7] = UInt8(udpChk & 0xFF)
        return p
    }
}

/// A parsed UDP datagram inside an IPv4 payload.
public struct UDPDatagram {
    public static let dnsPort = 53
    public let srcPort: Int
    public let dstPort: Int
    public let payload: [UInt8]

    public static func parse(_ ip: IPv4Packet) -> UDPDatagram? {
        if ip.proto != IPv4Packet.protoUDP { return nil }
        let off = ip.payloadOffset
        if ip.payloadLength < 8 { return nil }
        let buf = ip.raw
        let srcPort = (Int(buf[off]) << 8) | Int(buf[off + 1])
        let dstPort = (Int(buf[off + 2]) << 8) | Int(buf[off + 3])
        let udpLen = (Int(buf[off + 4]) << 8) | Int(buf[off + 5])
        let dataLen = max(0, min(udpLen - 8, ip.length - (off + 8)))
        return UDPDatagram(srcPort: srcPort, dstPort: dstPort,
                           payload: Array(buf[(off + 8)..<(off + 8 + dataLen)]))
    }
}

File: android/app/src/main/java/com/forrestlasiter/umbra/net/Ipv4Packet.kt
package com.forrestlasiter.umbra.net

/**
 * Just enough IPv4 to route DNS through the sinkhole: parse the header to find
 * UDP datagrams, and rebuild a reply packet (src/dst swapped) with a correct
 * header checksum. Not a general IP stack — the DNS-only route means only DNS
 * packets ever reach us.
 */
data class Ipv4Packet(
    val raw: ByteArray,
    val length: Int,
    val ihl: Int,             // header length in bytes
    val protocol: Int,        // 17 = UDP
    val srcAddr: ByteArray,   // 4 bytes
    val dstAddr: ByteArray,   // 4 bytes
) {
    val payloadOffset: Int get() = ihl
    val payloadLength: Int get() = length - ihl

    companion object {
        const val PROTO_UDP = 17

        fun parse(buf: ByteArray, length: Int): Ipv4Packet? {
            if (length < 20) return null
            val version = (buf[0].toInt() and 0xF0) ushr 4
            if (version != 4) return null
            val ihl = (buf[0].toInt() and 0x0F) * 4
            if (ihl < 20 || ihl > length) return null
            val protocol = buf[9].toInt() and 0xFF
            val src = buf.copyOfRange(12, 16)
            val dst = buf.copyOfRange(16, 20)
            return Ipv4Packet(buf, length, ihl, protocol, src, dst)
        }

        /**
         * Assemble a complete IPv4 + UDP packet. Recomputes the IPv4 header
         * checksum and the UDP checksum (pseudo-header included).
         */
        fun buildUdp(
            srcAddr: ByteArray, dstAddr: ByteArray,
            srcPort: Int, dstPort: Int, udpPayload: ByteArray,
        ): ByteArray {
            val udpLen = 8 + udpPayload.size
            val totalLen = 20 + udpLen
            val p = ByteArray(totalLen)

            // --- IPv4 header (20 bytes, no options) ---
            p[0] = 0x45.toByte()                       // version 4, IHL 5
            p[1] = 0                                    // DSCP/ECN
            p[2] = ((totalLen ushr 8) and 0xFF).toByte()
            p[3] = (totalLen and 0xFF).toByte()
            // id 0, flags 0, frag 0 (bytes 4..7 left zero)
            p[8] = 64                                   // TTL
            p[9] = PROTO_UDP.toByte()
            // header checksum bytes 10..11 left zero for the computation
            System.arraycopy(srcAddr, 0, p, 12, 4)
            System.arraycopy(dstAddr, 0, p, 16, 4)
            val ipChk = InternetChecksum.compute(p, 0, 20)
            p[10] = ((ipChk ushr 8) and 0xFF).toByte()
            p[11] = (ipChk and 0xFF).toByte()

            // --- UDP header (8 bytes) + payload ---
            val u = 20
            p[u] = ((srcPort ushr 8) and 0xFF).toByte()
            p[u + 1] = (srcPort and 0xFF).toByte()
            p[u + 2] = ((dstPort ushr 8) and 0xFF).toByte()
            p[u + 3] = (dstPort and 0xFF).toByte()
            p[u + 4] = ((udpLen ushr 8) and 0xFF).toByte()
            p[u + 5] = (udpLen and 0xFF).toByte()
            // checksum bytes 6..7 left zero, then filled below
            System.arraycopy(udpPayload, 0, p, u + 8, udpPayload.size)

            // UDP checksum over pseudo-header (src, dst, zero, proto, udpLen) + UDP.
            var seed = InternetChecksum.partialSum(srcAddr, 0, 4) +
                InternetChecksum.partialSum(dstAddr, 0, 4) +
                PROTO_UDP.toLong() + udpLen.toLong()
            var udpChk = InternetChecksum.compute(p, u, udpLen, seed)
            if (udpChk == 0) udpChk = 0xFFFF          // 0 means "no checksum"; use all-ones
            p[u + 6] = ((udpChk ushr 8) and 0xFF).toByte()
            p[u + 7] = (udpChk and 0xFF).toByte()
            return p
        }
    }
}
